File: 백준/Silver/1074. Z/Z.swift
import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let (n, r, c) = (input[0], input[1], input[2])

func z(_ n: Int, _ r: Int, _ c: Int) -> Int {
    let half = 1 << (n - 1)
    if n == 0 {
        return 0
    }
    if r < half && c < half {
        return z(n - 1, r, c)
    } else if r < half && c >= half {
        return half * half + z(n - 1, r, c - half)
    }  else if r >= half && c < half {
        return 2 * half * half + z(n - 1, r - half, c)
    } else {
        return 3 * half * half + z(n - 1, r - half, c - half)
    }
}

print(z(n, r, c))

File: 백준/Silver/1074. Z/Z.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

// Z
// https://www.acmicpc.net/problem/1074
public class Main {

	static int Z(int N, int r, int c) {
		int half = 1 << (N - 1);
		if (N == 0) {
			return 0;
		}
		if (r < half && c < half) {
			return Z(N - 1, r, c);
		} else if (r < half && c >= half) {
			return half * half + Z(N - 1, r, c - half);
		} else if (r >= half && c < half) {
			return 2 * half * half + Z(N - 1, r - half, c);
		} else {
			return 3 * half * half + Z(N - 1, r - half, c - half);
		}
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();

		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int r = Integer.parseInt(st.nextToken());
		int c = Integer.parseInt(st.nextToken());

		sb.append(Z(N, r, c));

		br.close();
		System.out.println(sb);
	}
}
